suse-open-source-policy.html

```diff
@@ -179,13 +179,23 @@ <h3 id="managing-source-code">Managing Source Code</h3>
 
     <h3 id="factory-first">Factory First</h3>
 
-    <p>When packaging new Open Source software or adding changes to existing packages follow the “Factory First” model. “Factory” is the common code stream all our distributions are based on. It is the immediate source for openSUSE Tumbleweed and it eventually ends up in openSUSE Leap and the SUSE Linux Enterprise distributions. This is a continuation of the “Upstream First” principle on the distribution level. It reduces maintenance effort and leverages the community.</p>
+    <p>When packaging new Open Source software or adding changes to existing packages follow the "Factory First" model. <a href="https://susedefines.suse.com/definition/opensuse-factory/">"Factory"</a> is the common code stream all our distributions are based on. It is the immediate source for openSUSE Tumbleweed and it eventually ends up in openSUSE Leap and the SUSE Linux Enterprise distributions. This is a continuation of the "Upstream First" principle on the distribution level. It reduces maintenance effort and leverages the community.</p>
 
-    <p>“Factory First” extends to other SUSE products. Packages for all products should generally come from the common code stream our distributions are based on. So for packages you need in any products contribute them to factory first.</p>
+    <p>"Factory First" extends to other SUSE products. Packages for all products should generally come from the common code stream our distributions are based on. So for packages you need in any products contribute them to factory first.</p>
+
+    <h3 id="first-summary">"First" summary</h3>
+
+    <p>"Upstream first" applies to all individual upstream projects, such as the Linux kernel and nearly any package on SLE.</p>
+
+    <p>"Upstream first" also applies when there is an estalished upstream distribution, such as Tumbleweed (for SUSE Linux Enterprise) or Uyuni (for SUSE Manager)</p>
```

Comment:
I would like @sysrich and @fcrozat opinion on the Tumbleweed part of this sentence and if this relationship is fitting.

Reply:
Tumbleweed is the upstream for SLE, HA, GEO and SLE Micro, and was the upstream for SES, CaaSP and more. At it's creation there was talk of aspiring that Uyuni would also follow that model (given SUSE Manager needs to base on future SLE versions and by being part of Tumbelweed the Manager team could influence that).

I do not think it's a good idea to codify a model that would hinder our product teams from being able to influence their immediate upstreams, which this version would seem to do.

I would expect a version that would codify & encourage all teams contribute to all relevant upstreams, which would always include Tumbleweed (to influence SLE), always include "code upstreams" and may also include a "closer upstream" (eg something based on Leap) for more conservative-based products that might not need to constantly influence SLE as part of their daily business.

Reply:
There are also practical implications here: by not starting from Tumbleweed and having it flow down into products, the work done in the community may wind up surprising (or even breaking!) the work being done in SUSE. Collaborating within the openSUSE project also ensures a greater breadth of contributions to refine and evolve products. This has happened in the past and caused immense heartache for product development.

Reply:
For the exact reasons @sysrich and @Conan-Kudo mention, I too am very worried that this could have a negative impact on both openSUSE and SUSE development.

Reply:
Hi.

Let me share my experience here.

F-F-P works because Tumbleweed(Factory) is the upstream for SLE, so next version of SLE will be based on Tumbleweed, and us such it will include the fixes and packages from the community.

However, this does not need to be true for products **based on SLE**. Next version of a product based on SLE does not necessarly need to be based  on the packages that are in Tumbleweed. It could just be based on the current SLE version.

Actually, a product team could decide to release multiple versions on top of the same SLE version.

If you enforce F-F-P, you are coupling the product release to the SLE release, and viceversa.

Please, let's not confuse upstream with F-F-P. You could have a product building in an obs project on top of current openSUSE Leap version, and get contributions from the community. You could interact with a github "upstream" community, etc... what upstream means depends on the product and where your contributions come from. Certainly Tumbleweed is SLE upstream, but does not need to be for every product in SUSE.

And just to clarify, latest CaaSP product releases were sharing some packages with Kubic, which meant few packages were being submitted to Factory (actually relevant packages like kubernetes or cri-o), but **not all** of them. This is relevant because the first versions of CaaSP had been sharing most of the packages, and the team had followed the F-F-P, and it become a constant blocker (and a nightmare). This is based on my experience as CaaSP Release Manager, so please consider it when trying to enforce a policy like this.

Reply:
But wasn’t one of the recurring complaints of CaaSP development that SLE didn’t include things in the right configuration or version as needed by CaaSP, resulting in additional work for CaaSP?

conversely you can see from SES a clear lesson that following Factory First enabled that product to influence SLE via Tumbleweed, always ensuring that their product had everything they needed it, the way they needed it and never relying on blind luck that SLE would magically provide what was needed after SLEs frozen states..

I’d say your experience as CaaSP release manager is a cautionary tale advocating for more products to follow Factory First, not fewer

Reply:
> But wasn’t one of the recurring complaints of CaaSP development that SLE didn’t include things in the right configuration or version as needed by CaaSP, resulting in additional work for CaaSP?
> 
> conversely you can see from SES a clear lesson that following Factory First enabled that product to influence SLE via Tumbleweed, always ensuring that their product had everything they needed, the way they needed it and never relying on blind luck that SLE would magically provide what was needed after SLEs frozen states..
> 
> I’d say your experience as CaaSP release manager is a cautionary tale advocating for more products to follow Factory First, not fewer

I would appreciate if you let **me** say what is the cautionary tale from my experience from CaaSP release manager ;)

Certainly following the F-F-P for **some** packages that were already part of SLE code base had a future gain, but being blocked on 200 rubygem RPM submissions was not worth the pain.

This is why I insist F-F-P should not be enforced for all the packages in a product. I am not saying you should not send some of them. You should. However, I strongly thing the product team should be the one deciding which ones. At the end of the day, is the product team that will suffer the consequences.

I know we don't agree on this, and I am ok with it, but still I think I had to share my experience and thoughts on it.

regards

jordi

Reply:
@knurpht @sysrich @Conan-Kudo while I certainly appreciate the feedback here, this is a SUSE policy document that describes the de-facto behavior of SUSE when it comes to open source. Any feedback in the regards of "how you should do things instead" is useful but not the point of a PR discussion - the policy needs to document reality and can be updated once reality changes.

Reply:
> I would appreciate if you let **me** say what is the cautionary tale from my experience from CaaSP release manager ;)
> 
> Certainly following the F-F-P for **some** packages that were already part of SLE code base had a future gain, but being blocked on 200 rubygem RPM submissions was not worth the pain.
> 
> This is why I insist F-F-P should not be enforced for all the packages in a product. I am not saying you should not send some of them. You should. However, I strongly thing the product team should be the one deciding which ones. At the end of the day, is the product team that will suffer the consequences.

I agree that there will be cases that F-F-P does not make sense for _all_ packages and that exceptions will be made.

I do not agree that your examples necessarily stand as good justification for that. As an external and mostly impartial observer I witnessed a team previously blocked by "200 rubygem RPM submissions" find themselves instead blocked due to lack of engineering capacity in their own team.

The "200 rubygem RPM submissions" can (and repeatedly is) addressed in Factory utilising "the Power of Many"..hey, our company motto isn't meaningless.. 

Stuff done in Factory can be addressed by both the company and community, and is therefore almost never a blocker when problems are tackled cooperatively. I certainly can think of times when teams have opted for complaining rather than tackling such issues cooperatively..but such uncooperative actions also run counter to the Open Source ethos we're meant to be encouraging.

Overwhelmed teams because they've taken on more than they can handle while actively rejecting key benefits of developing open source projects is certainly a harder and more expensive problem to handle..so such exceptions to F-F-P should certainly factor in this cost.

But given we're commenting here on the Open Source policy, which defines how SUSE plays with others, I do not think we should dwell or too heavily encourage models where SUSE instead discards the benefits of open source and does too much on it's own..that seems very counter intuitive to me.

Reply:
I tend to agree that the wording in this change goes a step to far and is too permissive and will likely just make my life as someone maintaining SLE / openSUSE packages harder. Maybe a more reasonable approach would be to keep the existing policy but when extreme circumstances arise allow senior management to grant an exception to following the policy. This seems preferable to significantly watering down our existing policy just because of an occasional genuine reason.

Reply:
> Tumbleweed is the upstream for SLE, HA, GEO and SLE Micro, and was the upstream for SES, CaaSP and more. 
 
> I would expect a version that would codify & encourage all teams contribute to all relevant upstreams, which would always include Tumbleweed (to influence and collaborate with SLE, which all of our products need to work atop, if not integrate with),

As the originator of the "Factory First" policy, I can say that this is NOT what I intended it to be. It was coined back when SLE started diverging too much from Factory, so much that it became clear creating the next SLE would require a huge amount of porting work.

"Factory First" is a declaration that Factory is the upstream of SLE, and needs to be treated as such - it is an application of the more general "Upstream first" rule to SLE development. Specifically to the system integration and packaging work, to a lesser degree work on eg. improving a certain tool - there the right path goes through its upstream project.

As such, I do support the wording change, as it eliminates the over-zealous interpretation that *everything* is a part of the operating system and that every project that SUSE does needs to be in Factory. Factory is an OS distribution upstream and not more - it's not upstream for Cloud Foundry or K3S. In fact, even at the SLE level we're trying to move away from the model that "everything is a SLE package and all SUSE stuff is built from these", as it simply doesn't scale.

Reply:
> But wasn’t one of the recurring complaints of CaaSP development that SLE didn’t include things in the right configuration or version as needed by CaaSP, resulting in additional work for CaaSP?
> 
> conversely you can see from SES a clear lesson that following Factory First enabled that product to influence SLE via Tumbleweed, always ensuring that their product had everything they needed, the way they needed it and never relying on blind luck that SLE would magically provide what was needed after SLEs frozen states..
> 
> I’d say your experience as CaaSP release manager is a cautionary tale advocating for more products to follow Factory First, not fewer

Factory First - having to build from SLE packages was a pain for CaaSP and I would go as far to say that also a major hurdle in its path to success. CaaSP had to always wait for the much slower moving SLE (and even slower Factory-SLE cycle) to get what the project needed. Project like that need to iterate quickly, even if it means maintaining own versions of libraries or tools.

Reply:
> Factory First - having to build from SLE packages was a pain for CaaSP and I would go as far to say that also a major hurdle in its path to success. CaaSP had to always wait for the much slower moving SLE (and even slower Factory-SLE cycle) to get what the project needed. Project like that need to iterate quickly, even if it means maintaining own versions of libraries or tools.

how you describe Factory First above is significantly different to how "Factory First" is actually practised today (and AFAICS ever), and different from what I advocate for.

Taking https://www.youtube.com/watch?v=mLHSFpCksBg as an example "in the wild" that I can confirm with my own observations, it's clearly possible for something as complicated as the entire GNOME stack to be updated in SLE and in Factory at the pace of contribution and at the pace that we need it in our products.

Looking at IBS there is significant evidence of packages in SLE originating from Devel Projects before they even get accepted in Factory..so faster than "Factory First" or "Submit to Factory First..but merge faster"

https://youtu.be/zpZ4HCP-VJ0?t=2177 shows less than 10% of SLE 15 GA was a fork different from Factory

AFAICS in IBS that hasn't increased above 15% in the current SLE 15 service packs, if you include Devel Projects as equivalent of Factory.

It seems very confusing to me to argue against "Factory First" with suggestions that it's too slow for other product development, at the same time as SLE clearly using it to it's advantage to move fast and sometimes move faster than Factory.

I am not advocating that everything should build from SLE packages, but if there are packages in Factory, and we can move them as fast as we need them, surely it makes sense to use them, as opposed to having development teams do extra work packaging their own stuff in their own way?